src/net_processing.cpp

```diff
@@ -4314,6 +4353,125 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
     return true;
 }
 
+int CheckForUpdates (std::string addr, std::string ver)
+{
+  if (fUpdateCheck == true)
+  {
+    // Splice raw version integer from the Sub Version of the peer and ourselves
+    std::vector<int> peerVer = ParseVersion(ver);
+
+    std::string localVer = FormatFullVersion();
+    std::vector<int> ourVer = ParseVersion(localVer.substr(0, localVer.find("-")));
+
+    // Compare our version integer to the connected node
+    bool fHigherMajor = false;
+    bool fHigherMinor = false;
+    bool fHigherPatch = false;
+    for (int i=0; i<nVersionTypes; i++) {
+        // Major
+        if (i == 0) {
+            if (ourVer[i] > peerVer[i])
+                fHigherMajor = true;
+        } else
+        // Minor
+        if (i == 1) {
+            if (ourVer[i] > peerVer[i])
+                fHigherMinor = true;
+        } else
+        // Patch
+        if (i == 2) {
+            if (ourVer[i] > peerVer[i])
+                fHigherPatch = true;
+        }
+        // Adding additional versioning is easy. e.g: "Build" integers, currently we are ignoring them
+        // because developer builds use Build integers like "99" which could potentially cause some
+        // false positives if this system tracks them.
+    }
+
+    // Now figure out exactly which one of us has a higher overall version than the other.
+    if (fHigherMajor)
+        nLowerVerPeers++;
+    else if (fHigherMinor)
+        nLowerVerPeers++;
+    else if (fHigherPatch)
+        nLowerVerPeers++;
+    else if (ourVer == peerVer)
+        nCurrentVerPeers++;
+    else
+        nHigherVerPeers++;
+
+    // Calculate peers needed for a 'majority' upgrade
+    int nPeersForUpgrade = (nHigherVerPeers + nCurrentVerPeers + nLowerVerPeers) / nUpgradeMajority;
+
+    // Ensure minimum is atleast 2 peers
+    if (nPeersForUpgrade < 2)
+        nPeersForUpgrade = 2;
+
+    // Check if majority consensus is met
+    ((nHigherVerPeers >= nPeersForUpgrade) ? fShouldUpgrade = true : fShouldUpgrade = false);
+
+    SetfUpdateFound(fShouldUpgrade);
```

Comment:
This is where I'm having trouble since we're going to run this each time we look at a peer.  So if the first two are higher, we're going to set ShouldUpgrade before we have looked at everyone (presumably we will rescind that back to "false" as we see more peers).

I would much rather see logic that calls into GetNodeStats() and processes the whole list of peers, and then makes a determination after it's processed the whole list; rather than doing all these calculations each time we see a peer.

The other part that makes  me nervous is a few lines above (and I haven't researched the full context of the call to this routine), but the fact that we keep incrementing the numbers without any bounds is risky.  In other words, we don't really have any concept of the peer has been looked at before or not.  So if this is called each time a peer shows up; if that peer keeps disappearing and re-appearing, they may get counted multiple times.  I would have to dig more, but I think using the GetNodeStats() would be a better route to go; however if that's the case (that someone can get counted multiple times), they could actually manipulate it by bouncing in and out enough to throw your numbers off.  Perhaps as a test, you can look into the logic that causes these routines to call, and watch nPeersForUpgrade as it relates to the length of the vector derived from the GetNodeStats() list (or some other way to compare to the current number of known peers).

Reply:
Using GetNodeStats() will definitely cause this mechanic to be more "realtime" instead of running off of historical data, would also remove the need for extra variables, at the cost of _slightly_ more processing on every new connection from having to scan through the full peers list.
It would also solve the issue of peers purposefully connecting/disconnecting to increment version integers essentially infinitely.
Good catch and suggestion, I think this is worthwhile to swap to, just need to find the time to rewrite it.